text/0000-std-aware-cargo.md

```diff
@@ -0,0 +1,262 @@
+
+This approach has [gathered support] from various [rust team members], and this RFC aims to take inspiration from tools and workflows like [xargo], and integrate them into Cargo itself.
+
+[xargo]: https://github.com/japaric/xargo
+[deprecated]: https://github.com/japaric/xargo/issues/193
+[gathered support]: https://github.com/japaric/xargo/issues/193#issuecomment-359180429
+[rust team members]: https://www.ncameron.org/blog/cargos-next-few-years/
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This proposal aims to make `core` and `std` feel a little bit less like a special case compared to other dependencies to the end users of Cargo. This will be achieved by using idioms used for interacting, configuring, modifying, and patching other crates in a similar method for `core` and `std`.
+
+This RFC proposes the following concrete changes, which may or may not be implemented in this order, and may be done incrementally. The details and caveats around these stages are discussed in the Reference Level Explanation.
+
+1. Allow developers of root crates to recompile `core` (and `compiler-builtins`) when their desired target does not match one available as a `rustup target add` target, without the usage of a nightly compiler. This version of `core` would be built from the same source files used to build the current version of `rustc`/`cargo`.
+2. Introduce the concept of "stable features" for `core`, which allow the end user to influence the behavior of their custom version of `core`, without the use of a nightly compiler.
+3. Extend the new behaviors described in step 1 and 2 for `std` (and `alloc`).
+4. Allow the user to provide their own custom source versions of `core` and `std`, allowing for deep customizations when necessary. This will require a nightly version of the compiler.
+
+As a new concept, the items above propose the existence of "stable features" for `core` and `std`. These features would be considered stable with the same degree of guarantees made for stability in the rest of the language. These features would allow configuration of certain functionalities of `core` or `std`, in a way decided at compile time.
+
+For example, we could propose a feature called `force-tiny-fmt`, which would use different algorithms to implement `fmt` for use on resource constrained systems. The developer of the root crate would be able to choose the default behavior, or the `force-tiny-fmt` behavior while still retaining the ability of using a stable compiler.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A reference-level explanation is made for each of the items enumerated above.
+
+## 1 - Allow developers of root crates to recompile `core`
+
+### Use Case
+
+For developers working with new targets not yet supported by the Rust project, this feature would allow the compilation of `core` for any target that can be specified as a valid target json format.
```

Comment:
Could you elaborate why json is chosen?

Reply:
The json format I mention is already used for custom target definitions, let me try to find the reference for this.

Reply:
This is the best documentation I've found for this, I'll add a link to the RFC: https://github.com/rust-lang/rust/blob/b7c6e8f1805cd8a4b0a1c1f22f17a89e9e2cea23/src/librustc_target/spec/mod.rs#L11-L45

Reply:
Also https://rust-lang.github.io/rfcs/0131-target-specification.html (mostly as a note to self)